wandb/sdk/launch/builder/docker_builder.py

```diff
@@ -0,0 +1,153 @@
+"""Implementation of the docker builder."""
+import logging
+import os
+from typing import Any, Dict
+
+import wandb
+import wandb.docker as docker
+from wandb.sdk.launch.builder.abstract import AbstractBuilder
+from wandb.sdk.launch.environment.abstract import AbstractEnvironment
+from wandb.sdk.launch.registry.abstract import AbstractRegistry
+
+from .._project_spec import (
+    EntryPoint,
+    LaunchProject,
+    create_metadata_file,
+    get_entry_point_command,
+)
+from ..registry.local_registry import LocalRegistry
+from ..utils import LOG_PREFIX, LaunchError, sanitize_wandb_api_key
+from .build import (
+    _create_docker_build_ctx,
+    generate_dockerfile,
+    validate_docker_installation,
+)
+
+_GENERATED_DOCKERFILE_NAME = "Dockerfile.wandb-autogenerated"
+_logger = logging.getLogger(__name__)
+
+
+class DockerBuilder(AbstractBuilder):
+    """Builds a docker image for a project.
+
+    Attributes:
+        builder_config (Dict[str, Any]): The builder config.
+
+    """
+
+    builder_type = "docker"
+    base_image = "python:3.8"
+    target_platform = "linux/amd64"
+
+    def __init__(
+        self,
+        environment: AbstractEnvironment,
+        registry: AbstractRegistry,
+        verify: bool = True,
+        login: bool = True,
+    ):
+        """Initialize a DockerBuilder.
+
+        Args:
+            environment (AbstractEnvironment): The environment to use.
+            registry (AbstractRegistry): The registry to use.
+            verify (bool, optional): Whether to verify the functionality of the builder.
+            login (bool, optional): Whether to login to the registry.
+
+        Raises:
+            LaunchError: If docker is not installed
+        """
+        self.environment = environment  # Docker builder doesn't actually use this.
```

Comment:
Yeah this is sort of weird. The registry and builder are both configured with an environment. The registry uses its environment to get push/pull credentials. The builder uses its environment to access storage to upload and download build contexts. The docker build doesn't need to copy or pull any build contents since the build is local. But the docker builder can be used with a remote registry, in which case the registry uses the environment to get push/pull secrets.

Reply:
Yeah that makes sense.